glance_store/_drivers/s3.py

```diff
@@ -691,127 +787,135 @@ def delete(self, location, context=None):
-    from boto.exception import S3ResponseError
+def create_bucket_if_missing(s3_client, bucket_id, conf):
+
     try:
-        s3_conn.get_bucket(bucket)
-    except S3ResponseError as e:
-        if e.status == http_client.NOT_FOUND:
+        s3_client.head_bucket(Bucket=bucket_id)
+    except ClientError as e:
+        error_code = e.response['Error']['Code']
+        if error_code == '404':
             if conf.glance_store.s3_store_create_bucket_on_put:
-                host = conf.glance_store.s3_store_host
-                location = get_s3_location(host)
                 try:
-                    s3_conn.create_bucket(bucket, location=location)
-                except S3ResponseError as e:
-                    msg = (_("Failed to add bucket to S3.\n"
-                             "Got error from S3: %s.") %
-                           encodeutils.exception_to_unicode(e))
+                    host = conf.glance_store.s3_store_host
+                    location = get_s3_location(host)
+                    s3_client.create_bucket(Bucket=bucket_id,
+                                            CreateBucketConfiguration={'LocationConstraint': location})
+                except ClientError as e:
+                    msg = (_("""
+                             Failed to add bucket to S3.
+                             Got error from S3: %s.
+                             """) % encodeutils.exception_to_unicode(e))
                     raise glance_store.BackendException(msg)
             else:
-                msg = (_("The bucket %(bucket)s does not exist in "
-                         "S3. Please set the "
-                         "s3_store_create_bucket_on_put option "
-                         "to add bucket to S3 automatically.")
-                       % {'bucket': bucket})
+                msg = (_("""
+                         The bucket %(bucket_id)s does not exist in
+                         S3. Please set the
+                         s3_store_create_bucket_on_put option
+                         to add bucket to S3 automatically.
+                         """) % {'bucket_id': bucket_id})
                 raise glance_store.BackendException(msg)
+        else:
+            msg = (_("""Got error from S3: %s.""") % encodeutils.exception_to_unicode(e))
+            raise glance_store.BackendException(msg)
 
 
-def get_key(bucket, obj):
-    """
-    Get a key from a bucket
-    :param bucket: The ``boto.s3.Bucket``
-    :param obj: Object to get the key for
-    :raises: ``glance_store.exceptions.NotFound`` if key is not found.
-    """
-
-    key = bucket.get_key(obj)
-    if not key or not key.exists():
-        msg = (_("Could not find key %(obj)s in bucket %(bucket)s") %
-               {'obj': obj, 'bucket': bucket})
-        LOG.debug(msg)
-        raise exceptions.NotFound(message=msg)
-    return key
-
+def get_key(s3_client, bucket_id, object_key):
+    try:
+        key = s3_client.get_object(Bucket=bucket_id, Key=object_key)
```

Comment:
以下のように1行で書くか、returnはtryブロックの外に出すかしてtryブロックを小さくした方が良いと思います。
```suggestion
        return s3_client.get_object(Bucket=bucket_id, Key=object_key)
```